// @sendbird/react-uikit-message-template-view/dist/util/index.js
import { ComponentType, FlexSizeSpecValue } from '@sendbird/uikit-message-template';
export function isWrappedText(view) {
    var _a;
    return view.type === ComponentType.Text && ((_a = view.width) === null || _a === void 0 ? void 0 : _a.value) == FlexSizeSpecValue.WrapContent;
}
// Compares the size spec to the option and returns whether they are equal or not.
export const isSizeCompatibleWithOption = ({ size, option, }) => {
    if (option === 'wrap') {
        return size.type === 'flex' && size.value == FlexSizeSpecValue.WrapContent;
    }
    if (option === 'fill') {
        return size.type === 'flex' && size.value == FlexSizeSpecValue.FillParent;
    }
    return size.type === 'fixed';
};
export const isGif = (url) => {
    var _a;
    if (typeof url !== 'string')
        return false;
    const extension = (_a = url.split('.').pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
    return extension === 'gif';
};
export const isNumber = (val) => {
    if (typeof val === 'string') {
        return !Number.isNaN(Number(val));
    }
    return typeof val === 'number' && !Number.isNaN(val);
};
export const clx = (...names) => {
    return names.filter((it) => !!it).join(' ');
};
export const hasValidURLProtocol = (url = '') => {
    return ['http://', 'https://', 'ftp://'].some((protocol) => url.startsWith(protocol));
};
/**
 * @param url - url to be checked
 * @returns url with http:// protocol if it doesn't have any protocol
 * @example
 * returnUrl('www.sendbird.com') // returns 'http://www.sendbird.com'
 * returnUrl('https://www.sendbird.com') // returns 'https://www.sendbird.com'
 * returnUrl('ftp://www.sendbird.com') // returns 'ftp://www.sendbird.com'
 * returnUrl('sendbird.com') // returns 'https://sendbird.com'
 **/
export const getValidURL = (url = '') => {
    if (hasValidURLProtocol(url))
        return url;
    return `https://${url}`;
};
